crates/walrus-sdk/src/client/quilt_client.rs

```diff
@@ -0,0 +1,583 @@
+                metadata,
+                &sliver_indices,
+                certified_epoch,
+                Self::MAX_RETRIEVE_SLIVERS_ATTEMPTS,
+                Self::TIMEOUT_DURATION,
+            )
+            .await;
+
+        if let Ok(slivers) = retrieved_slivers {
+            let sliver_refs: Vec<_> = slivers.iter().collect();
+            let decoder = V::QuiltConfig::get_decoder_with_quilt_index(&sliver_refs, index);
+            identifiers
+                .iter()
+                .map(|identifier| {
+                    decoder
+                        .get_blob_by_identifier(identifier)
+                        .map_err(ClientError::other)
+                })
+                .collect::<Result<Vec<_>, _>>()
+        } else {
+            let quilt = self.get_quilt_enum(metadata, certified_epoch).await?;
+            identifiers
+                .iter()
+                .map(|identifier| {
+                    quilt
+                        .get_blob_by_identifier(identifier)
+                        .map_err(ClientError::other)
+                })
+                .collect::<Result<Vec<_>, _>>()
+        }
+    }
+
+    /// Retrieves the quilt from Walrus.
+    async fn get_quilt_enum(
+        &self,
+        metadata: &VerifiedBlobMetadataWithId,
+        certified_epoch: Epoch,
+    ) -> ClientResult<QuiltEnum> {
+        let quilt = self
+            .client
+            .request_slivers_and_decode::<Primary>(certified_epoch, metadata)
+            .await?;
+        let encoding_config_enum = self
+            .client
+            .encoding_config()
+            .get_for_type(metadata.metadata().encoding_type());
+
+        QuiltEnum::new(quilt, &encoding_config_enum).map_err(ClientError::other)
+    }
+}
+
+/// Stores quilts.
+impl QuiltClient<'_, SuiContractClient> {
+    /// Constructs a quilt from a list of blobs.
+    pub async fn construct_quilt<V: QuiltVersion>(
+        &self,
+        blobs: &[QuiltStoreBlob<'_>],
+        encoding_type: EncodingType,
+    ) -> ClientResult<V::Quilt> {
+        let encoder = V::QuiltConfig::get_encoder(
+            self.client.encoding_config().get_for_type(encoding_type),
+            blobs,
+        );
+
+        encoder.construct_quilt().map_err(ClientError::other)
+    }
+
+    /// Converts a list of blobs with paths to a list of [`QuiltStoreBlob`]s.
+    ///
+    /// The on-disk file names are used as identifiers for the quilt patches.
+    /// If the file name is not valid UTF-8, it will be replaced with "unnamed-blob-<index>".
+    fn assign_identifiers_with_paths(
+        blobs_with_paths: &[(PathBuf, Vec<u8>)],
+    ) -> Vec<QuiltStoreBlob> {
+        blobs_with_paths
+            .iter()
+            .enumerate()
+            .map(|(i, (path, blob))| {
+                QuiltStoreBlob::new(
+                    blob,
+                    path.file_name()
+                        .and_then(|file_name| file_name.to_str())
```

Comment:
Important: We should have some way of handling duplicate filenames. 

One idea would be to use relative paths to some specified path (we have already deduplicated the file paths).